### DISCUSSION TITLE
Create REST API for custom component using Joomla 4.0

### DISCUSSION THREAD

**Discussion by ahamed** [joomla/joomla-cms#35282] [Q&A] [answered]
Hello Joomlars!
I've noticed that Joomla! 4 brings so many API endpoints for its core components. How do I create REST API for a custom component. Is there any documentation for that?

Thanks.

**Comment by alikon** [comment, marked as answer]
https://docs.joomla.org/J4.x:Adding_an_API_to_a_Joomla_Component

**Reply by ahamed** [reply]
Thank you.

**Reply by quangthan155** [reply]
Dear bro! could you help me to get all featured articles ? Or getting all articles by tags, Do we have to code more? All what i mean is to get by API. 
Looking forward to your help!

**Comment by cappuccinonet** [comment]
Is it be possible to just create a Plugin - without the component part, too?
I. e. like [com_api](https://docs.techjoomla.com/com_api/) is doing it.

**Reply by alexandreelise** [reply]
Hi @cappuccinonet, Joomla! 4 is working almost the same way. You can create a plugin of type webservices (the webservices plugin folder) and you create a plugin with the custom routes almost the same as com_api and plugins concept named apps in techjoomla.
Here is a blog article showing an example webservice plugin [Click here to see the blog article](https://alexapi.cloud/en/blog-joomla-specialist/helloapi-joomla4-component-webservices)

Like @alikon mentioned there is an official article on that too

Have a great day super joomler! Happy coding!

**Reply by laoneo** [reply]
This answer here is wrong as @cappuccinonet asked for a tutorial without a component. As routes are controller based, basically you have to do register a controller in the plugin which gets called from the route. Or you create your own route implementation which returns the controller.

**Reply by alexandreelise** [reply]
@laoneo I didn't express it like you said but that's exactly what I meant. Thanks for the clarification Allon.

**Reply by Node0** [reply]
I'm working on tooling i.e. `componentMaker.py` (a python script that takes some CLI parameters and creates a custom (intentionally simplified) component skeleton.
I'm also working on a `pluginMaker.py` script which initially will be intended to create webservices plugins, but there's no reason it couldn't be extended to take the plugin type as input from the user (from the CLI as part of normal config params) and then pop out one of any of the 22 custom types of Joomla plugins, I'm just not focused on comprehensive tooling support right now as my focus at the moment requires API capabilities.

Here's the repo if you're interested: https://github.com/Node0/joomla-tools

**Comment by cappuccinonet** [comment]
@alexandreelise Thanks for the hint. Looks a bit complex to create a component as well. In my Rest API I just need to connect to existing (non-Joomla) database tables and output the query  results as JSON. - Is there no easier way to achieve this than writing a component as well as a plugin?

**Reply by Node0** [reply]
Hi @cappuccinonet Well if it makes you feel any better, I'm in the same boat and have begun tooling to create scaffolding for custom components have a look at my joomla-tools repo that will do most of the work for you in creating a component skeleton, then you just need to create the api folder and setup the api controller (in addition to the webservices plugin of course) but this way you're only fiddling with 5 to 6 files instead of 18 to 20.

Here's my repo: https://github.com/Node0/joomla-tools

**Comment by MarcelSchuermann** [comment]
Is there already a new working version for com_api available or a new solution for a general custom rest api endpoint?
See also discussion on com_api discussion:
https://github.com/techjoomla/com_api/issues/154

What could be the best approach to migrate an existing com_api solution (including the existing api keys, api tables and data as well as adjusted component files) to the new J4 API-Architecture?

**Reply by alexandreelise** [reply]
@cappuccinonet Here is what another Super Joomler did with the latest up-to-date Joomla APIs and Web Services 

https://developingextensionsforjoomla5.com/blog/a-comprehensive-workshop-on-integrating-web-services-into-your-components-at-joomladay-france

It's not an affiliation link. I'm just happy to share what other fellow programmers do.

Here is a ready to use repo to he hid with a working example ready to use 

https://github.com/carcam/webservices-jd24fr

Hope it helps you more. Enjoy and happy new year.

One more thing... If all of you in this conversation need a comprehensive documentation here is a Postman documentation link with more than a hundred endpoint of all Joomla Web Services even those that are still in early stage not published officially yet 
here it is. Happy new year gift to all of you. With this you can generate your code in any language such as python for @Node0 , PHP , curl, Java, Go, Swift, etc..and many more for others. Here it is
https://documenter.getpostman.com/view/7443864/2s9YywfKez

**Reply by alexandreelise** [reply]
@quangthan155 Here is what you might be looking for https://documenter.getpostman.com/view/7443864/2s9YywfKez
